cdap-app-templates/cdap-dq/src/main/java/co/cask/cdap/dq/DataQualityApp.java

```diff
@@ -0,0 +1,326 @@
+   * Generate and write an aggregation with the data collected by the mapper
+   */
+  public static class AggregationReducer extends Reducer<Text, DataQualityWritable, byte[], Put>
+    implements ProgramLifecycle<MapReduceContext> {
+    private static final Gson GSON = new Gson();
+    private String sourceId;
+    private Map<String, Set<String>> fieldAggregations;
+    long timeKey = 0;
+
+    @Override
+    public void initialize(MapReduceContext mapReduceContext) throws Exception {
+      timeKey = mapReduceContext.getLogicalStartTime();
+      sourceId = mapReduceContext.getSpecification().getProperties().get("sourceId");
+      fieldAggregations = GSON.fromJson(mapReduceContext
+                                          .getSpecification().getProperties().get("fieldAggregations"),
+                                        TOKEN_TYPE_MAP_STRING_SET_STRING);
+    }
+
+    @Override
+    public void reduce(Text key, Iterable<DataQualityWritable> values, Context context)
+      throws IOException, InterruptedException {
+      LOG.debug("timestamp: {}", timeKey);
+      Set<String> aggregationTypesSet = fieldAggregations.get(key.toString());
+      List<AggregationTypeValue> aggregationTypeValueList = new ArrayList<>();
+      AggregationsRowKey aggregationsRowKey = new AggregationsRowKey(timeKey, sourceId);
+      byte[] fieldColumnKey = Bytes.toBytes(key.toString());
+      for (String aggregationType : aggregationTypesSet) {
+        boolean isCombinable = true;
+        try {
+          Class<?> aggregationClass = Class.forName("co.cask.cdap.dq.functions." + aggregationType);
+          BasicAggregationFunction instance = (BasicAggregationFunction) aggregationClass.newInstance();
+          isCombinable = instance instanceof CombinableAggregationFunction;
+          for (DataQualityWritable value : values) {
+            instance.add(value);
+          }
+          ValuesRowKey valuesRowKey = new ValuesRowKey(timeKey, key.toString(), sourceId);
+          context.write(valuesRowKey.getTableRowKey(), new Put(valuesRowKey.getTableRowKey(),
+                                                               Bytes.toBytes(aggregationType),
+                                                               instance.aggregate()));
+
+          AggregationTypeValue aggregationTypeValue = new AggregationTypeValue(aggregationType, isCombinable);
+          aggregationTypeValueList.add(aggregationTypeValue);
+        } catch (ClassNotFoundException | RuntimeException | InstantiationException | IllegalAccessException e) {
+          throw new RuntimeException(e);
+        }
+      }
+      byte[] aggregationTypeListBytes = Bytes.toBytes(GSON.toJson(aggregationTypeValueList));
+      context.write(aggregationsRowKey.getTableRowKey(),
+                    new Put(aggregationsRowKey.getTableRowKey(), fieldColumnKey, aggregationTypeListBytes));
+      //TODO: There are issues if multiple apps use the same source and same fields, but different aggregations
```

Comment:
it is still an issue. I opened https://issues.cask.co/browse/CDAP-3538 which you can include in the comment